gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,199 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.ApiException;
+import com.google.api.gax.rpc.StatusCode.Code;
+import com.google.common.base.MoreObjects;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * Keeps the statistics about failed operations(both at RPC and ElementT) in {@link Batcher}. This
+ * provides the count of individual exception failure and count of each failed {@link Code} occurred
+ * in the batching process.
+ */
+class BatcherStats {
+
+  private final Map<Class, Integer> requestExceptionCounts = new HashMap<>();
+  private final Map<Code, Integer> requestStatusCounts = new HashMap<>();
+  private final Map<Class, Integer> entryExceptionCounts = new HashMap<>();
+  private final Map<Code, Integer> entryStatusCounts = new HashMap<>();
+  private final Object lock = new Object();
+  private int partialBatchFailures;
+
+  /**
+   * Records the count of the exception and it's type when complete batch is failed to apply.
+   *
+   * <p>Note: This method aggregates all the subclasses of {@link ApiException} under ApiException
+   * using the {@link Code status codes} and its number of occurrences.
+   */
+  void recordBatchFailure(Throwable throwable) {
+    Class exceptionClass = throwable.getClass();
+
+    synchronized (lock) {
+      if (throwable instanceof ApiException) {
+        Code code = ((ApiException) throwable).getStatusCode().getCode();
+        exceptionClass = ApiException.class;
+
+        int oldCount = MoreObjects.firstNonNull(requestStatusCounts.get(code), 0);
+        requestStatusCounts.put(code, oldCount + 1);
+      }
+
+      int oldExCount = MoreObjects.firstNonNull(requestExceptionCounts.get(exceptionClass), 0);
+      requestExceptionCounts.put(exceptionClass, oldExCount + 1);
+    }
+  }
+
+  /**
+   * Records partial failure occurred within per batch. For any exception within a batch, the {@link
+   * #partialBatchFailures} is incremented once. It also keeps the records of the count and type of
+   * each entry failure as well.
```

Comment:
Records partial failures within each batch. partialBatchFailures counts the number of batches that have at least one failed entry while entryStatusCounts and entryExceptionCounts track the count of the entries themselves.

Reply:
Thanks a lot for rewording this Javadoc.